analyze.go

```diff
@@ -94,13 +94,25 @@ func enterBlockCtx(stmt *Directive, ctx blockCtx) blockCtx {
 
 //nolint:gocyclo,funlen,gocognit
 func analyze(fname string, stmt *Directive, term string, ctx blockCtx, options *ParseOptions) error {
-	masks, knownDirective := directives[stmt.Directive]
-	currCtx, knownContext := contexts[ctx.key()]
+	masks := []uint{}
+	knownDirective := false
 
-	if !knownDirective {
-		for _, matchFn := range options.MatchFuncs {
-			if masks, knownDirective = matchFn(stmt.Directive); knownDirective {
-				break
+	currCtx, knownContext := contexts[ctx.key()]
+	directiveName := stmt.Directive
+	if len(options.DirectiveSources) == 0 {
+		masks, knownDirective = directives[directiveName]
+		if !knownDirective {
+			for _, matchFn := range options.MatchFuncs {
+				if masks, knownDirective = matchFn(stmt.Directive); knownDirective {
+					break
+				}
+			}
+		}
+	} else {
+		for _, matchFn := range options.DirectiveSources {
+			if masksInFn, found := matchFn(directiveName); found {
+				masks = append(masks, masksInFn...)
+				knownDirective = true
 			}
 		}
 	}
```

Comment:
If possible, can we simplify this by eliminating some indentation making it easier to read? I'm not sure that the following suggestion works as I haven't tried it and am coding in the comment editor.

```suggestion
	for _, matchFn := range options.DirectiveSources {
		if masksInFn, found := matchFn(directiveName); found {
			masks = append(masks, masksInFn...)
			knownDirective = true
		}
	}
	
	// This could be else if, but then I think you need to document that MatchFuncs
	// will be ignored if DirectiveSources is specified.
	if !knownDirective {
		masks, knownDirective = directives[directiveName]
		for _, matchFn := range options.MatchFuncs {
			if masks, knownDirective = matchFn(stmt.Directive); knownDirective {
				break
			}
		}
	}
```

Reply:
Cool! This refactor also indicates we prioritize `options.DirectiveSources`. Done in the latest commit